src/qibo/quantum_info/entropies.py

```diff
@@ -0,0 +1,924 @@
+"""Submodule with entropy measures."""
+
+from typing import Union
+
+import numpy as np
+from scipy.linalg import fractional_matrix_power
+
+from qibo.backends import GlobalBackend
+from qibo.config import PRECISION_TOL, raise_error
+from qibo.quantum_info.metrics import _check_hermitian_or_not_gpu, purity
+
+
+def shannon_entropy(prob_dist, base: float = 2, backend=None):
+    """Calculate the Shannon entropy of a probability array :math:`\\mathbf{p}`, which is given by
+
+    .. math::
+        H(\\mathbf{p}) = - \\sum_{k = 0}^{d^{2} - 1} \\, p_{k} \\, \\log_{b}(p_{k}) \\, ,
+
+    where :math:`d = \\text{dim}(\\mathcal{H})` is the dimension of the
+    Hilbert space :math:`\\mathcal{H}`, :math:`b` is the log base (default 2),
+    and :math:`0 \\log_{b}(0) \\equiv 0`.
+
+    Args:
+        prob_dist (ndarray or list): a probability array :math:`\\mathbf{p}`.
+        base (float): the base of the log. Defaults to  :math:`2`.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): backend to be used
+            in the execution. If ``None``, it uses :class:`qibo.backends.GlobalBackend`.
+            Defaults to ``None``.
+
+    Returns:
+        (float): Shannon entropy :math:`H(\\mathcal{p})`.
+    """
+    if backend is None:  # pragma: no cover
+        backend = GlobalBackend()
+
+    if isinstance(prob_dist, list):
+        # np.float64 is necessary instead of native float because of tensorflow
+        prob_dist = backend.cast(prob_dist, dtype=np.float64)
+
+    if base <= 0:
+        raise_error(ValueError, "log base must be non-negative.")
+
+    if len(prob_dist.shape) != 1:
+        raise_error(
+            TypeError,
+            f"Probability array must have dims (k,) but it has {prob_dist.shape}.",
+        )
+
+    if len(prob_dist) == 0:
+        raise_error(TypeError, "Empty array.")
+
+    if any(prob_dist < 0) or any(prob_dist > 1.0):
+        raise_error(
+            ValueError,
+            "All elements of the probability array must be between 0. and 1..",
+        )
+
+    if np.abs(np.sum(prob_dist) - 1.0) > PRECISION_TOL:
+        raise_error(ValueError, "Probability array must sum to 1.")
+
+    if base == 2:
+        log_prob = np.where(prob_dist != 0.0, np.log2(prob_dist), 0.0)
+    elif base == 10:
+        log_prob = np.where(prob_dist != 0, np.log10(prob_dist), 0.0)
+    elif base == np.e:
+        log_prob = np.where(prob_dist != 0, np.log(prob_dist), 0.0)
+    else:
+        log_prob = np.where(prob_dist != 0, np.log(prob_dist) / np.log(base), 0.0)
```

Comment:
some parts of this appear to be very similar in all the entropies, any chance we can delegate this to a common helper function?

Reply:
Done